emmarin/cl/cl/src/cl/indexed.rs

```diff
@@ -0,0 +1,325 @@
+            low_nf_paths.push(merkle::path(&self.hashed_leaves(), *idx));
+            self.leaves[*idx].next_value = *values.first().unwrap();
+            for w in values.windows(2) {
+                let prev = w[0];
+                let next = w[1];
+                new_leaves.push(Leaf {
+                    value: prev,
+                    next_value: next,
+                });
+            }
+            new_leaves.push(Leaf {
+                value: *values.last().unwrap(),
+                next_value: low_nf.next_value,
+            });
+        }
+
+        let mut mmr = MMR::new();
+        for leaf in &self.leaves {
+            mmr.push(&leaf.to_bytes());
+        }
+        assert_eq!(self.root(), frontier_root(&mmr.roots));
+
+        for new_leaf in new_leaves {
+            self.leaves.push(new_leaf);
+        }
+
+        BatchUpdateProof {
+            values,
+            low_nfs,
+            low_nf_paths,
+            mmr,
+        }
+    }
+}
+
+#[derive(Clone, Debug, Copy, Serialize, Deserialize)]
+pub struct Leaf {
+    value: Nullifier,
+    next_value: Nullifier,
+}
+
+impl Leaf {
+    pub fn to_bytes(&self) -> Vec<u8> {
+        self.value
+            .0
+            .into_iter()
+            .chain(self.next_value.0.into_iter())
+            .collect()
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct UpdateProof {
+    value: Nullifier,
+    low_nf_path: Path,
+    low_nf: Leaf,
+    mmr: MMR,
+}
+
+impl UpdateProof {
+    pub fn verify(&self, old_root: [u8; 32]) -> [u8; 32] {
+        assert!(in_interval(self.low_nf, self.value));
+
+        assert_eq!(
+            merkle::path_root(leaf(&self.low_nf.to_bytes()), &self.low_nf_path),
+            old_root
+        );
+
+        let new_leaf = Leaf {
+            value: self.value,
+            next_value: self.low_nf.next_value,
+        };
+
+        let mut updated_low_nf = self.low_nf;
+        updated_low_nf.next_value = self.value;
+
+        let updated_root = merkle::path_root(leaf(&updated_low_nf.to_bytes()), &self.low_nf_path);
+        assert_eq!(updated_root, frontier_root(&self.mmr.roots));
+
+        let mut mmr = self.mmr.clone();
+        mmr.push(&new_leaf.to_bytes());
+
+        frontier_root(&mmr.roots)
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct BatchUpdateProof {
+    values: Vec<Nullifier>,
+    low_nfs: Vec<Leaf>,
+    low_nf_paths: Vec<Path>,
+    mmr: MMR,
+}
+
+impl BatchUpdateProof {
+    pub fn nullifiers(&self) -> &[Nullifier] {
+        &self.values
+    }
+
+    pub fn verify(&self, old_root: [u8; 32]) -> [u8; 32] {
+        // TODO: old compiter in risc0, use is_sorted
+        for window in self.values.windows(2) {
+            assert!(window[0] < window[1]);
+        }
+
+        let mut new_leaves = Vec::new();
+        let mut cur_root = old_root;
+        let mut values = self.values.iter();
+
+        for (low_nf, path) in self.low_nfs.iter().zip(&self.low_nf_paths) {
```

Comment:
I don't think this works if you have multiple nullifiers in a batch that fit within the same interval.

What do you think of restructuring the batch update proof so that it's one nullifier batch per interval, and then take a vec of these interval batches to get batch updates across multiple leaves:

```
pub struct IntervalUpdateProof {
    low_nf: Leaf,
    low_nf_path: Path,
    mmr: MMR,
    new_nfs: Vec<Nullifier>,
}
struct BatchUpdateProof {
    intervals: Vec<IntervalUpdateProof>
}
```

Reply:
I think it's equivalent?